File: examples/python/async/create_raw_order.py
from BinaryOptionsToolsV2.pocketoption import PocketOptionAsync
from BinaryOptionsToolsV2.validator import Validator
from datetime import timedelta

import asyncio


async def main(ssid: str):
    # Initialize the API client
    api = PocketOptionAsync(ssid)
    await asyncio.sleep(5)  # Wait for connection to establish

    # Basic raw order example
    try:
        validator = Validator.contains('"status":"success"')
        response = await api.create_raw_order('42["signals/subscribe"]', validator)
        print(f"Basic raw order response: {response}")
    except Exception as e:
        print(f"Basic raw order failed: {e}")

    # Raw order with timeout example
    try:
        validator = Validator.regex(r'{"type":"signal","data":.*}')
        response = await api.create_raw_order_with_timout(
            '42["signals/load"]', validator, timeout=timedelta(seconds=5)
        )
        print(f"Raw order with timeout response: {response}")
    except TimeoutError:
        print("Order timed out after 5 seconds")
    except Exception as e:
        print(f"Order with timeout failed: {e}")

    # Raw order with timeout and retry example
    try:
        # Create a validator that checks for both conditions
        validator = Validator.all(
            [
                Validator.contains('"type":"trade"'),
                Validator.contains('"status":"completed"'),
            ]
        )

        response = await api.create_raw_order_with_timeout_and_retry(
            '42["trade/subscribe"]', validator, timeout=timedelta(seconds=10)
        )
        print(f"Raw order with retry response: {response}")
    except Exception as e:
        print(f"Order with retry failed: {e}")


if __name__ == "__main__":
    ssid = input("Please enter your ssid: ")
    asyncio.run(main(ssid))

File: examples/javascript/create_raw_order.js
const { PocketOption } = require("./binary-options-tools.node");
const { Validator } = require("./binary-options-tools.node");

async function main(ssid) {
  // Initialize the API client
  const api = new PocketOption(ssid);

  // Wait for connection to establish
  await new Promise((resolve) => setTimeout(resolve, 5000));

  // Basic raw order example
  try {
    const basicValidator = Validator.contains('"status":"success"');
    const basicResponse = await api.createRawOrder(
      '42["signals/subscribe"]',
      basicValidator,
    );
    console.log(`Basic raw order response: ${basicResponse}`);
  } catch (error) {
    console.log(`Basic raw order failed: ${error}`);
  }

  // Raw order with timeout example
  try {
    const timeoutValidator = Validator.regex(
      /{\"type\":\"signal\",\"data\":.*}/,
    );
    const timeoutResponse = await api.createRawOrderWithTimeout(
      '42["signals/load"]',
      timeoutValidator,
      { timeout: 5000 }, // 5 seconds
    );
    console.log(`Raw order with timeout response: ${timeoutResponse}`);
  } catch (error) {
    if (error.name === "TimeoutError") {
      console.log("Order timed out after 5 seconds");
    } else {
      console.log(`Order with timeout failed: ${error}`);
    }
  }

  // Raw order with timeout and retry example
  try {
    const retryValidator = Validator.all([
      Validator.contains('"type":"trade"'),
      Validator.contains('"status":"completed"'),
    ]);

    const retryResponse = await api.createRawOrderWithTimeoutAndRetry(
      '42["trade/subscribe"]',
      retryValidator,
      { timeout: 10000 }, // 10 seconds
    );
    console.log(`Raw order with retry response: ${retryResponse}`);
  } catch (error) {
    console.log(`Order with retry failed: ${error}`);
  }
}

// Check if ssid is provided as command line argument
const ssid = "";

main(ssid).catch(console.error);
